File: R/recode.R
#' Decode EAVS missing-value codes
#'
#' EAVS marks non-substantive responses with negative sentinel codes. From
#' 2018 on these are `-88` ("does not apply"), `-99` ("data not available"),
#' and `-77` ("valid skip"); in 2016 and earlier they are `-888888` and
#' `-999999`. This finds the numeric item columns in a raw EAVS table (see
#' [eavs_read()]), converts them to numbers, and sets sentinel values to `NA`.
#'
#' Identifier and text columns—FIPS codes, jurisdiction and state names,
#' comment and write-in fields—are left untouched. A column is treated as a
#' numeric item when, ignoring blanks and text placeholders, all of its values
#' are numbers. Because every EAVS item is a count or a rate, any negative
#' value in an item column is read as missing; recognized sentinels keep their
#' specific meaning, which [eavs_missing_status()] reports.
#'
#' @param data A data frame from [eavs_read()], or any table of EAVS columns.
#'
#' @return `data` with item columns converted to numeric and sentinel values
#'   set to `NA`. Other columns are unchanged.
#' @seealso [eavs_missing_status()] to see *why* each value is missing.
#' @export
#'
#' @examples
#' df <- data.frame(
#'   FIPSCode = c("01001", "01003"),
#'   A1a = c("1200", "-99"),
#'   A3a = c("15", "-88")
#' )
#' eavs_recode_missing(df)
eavs_recode_missing <- function(data) {
  recode_core(data)$values
}

#' Report why EAVS values are missing
#'
#' Companion to [eavs_recode_missing()]. Returns a table the same shape as
#' `data` in which each item value is replaced by the reason it is, or is not,
#' missing: `"reported"`, `"does_not_apply"`, `"not_available"`,
#' `"valid_skip"`, `"other_missing"`, or `"blank"`. This keeps the distinction
#' between "does not apply" and "data not available" that recoding to `NA`
#' alone would lose.
#'
#' @inheritParams eavs_recode_missing
#'
#' @return `data` with item columns replaced by an ordered factor of status
#'   labels. Identifier and text columns are unchanged.
#' @seealso [eavs_recode_missing()] for the recoded values themselves.
#' @export
#'
#' @examples
#' df <- data.frame(
#'   FIPSCode = c("01001", "01003"),
#'   A1a = c("1200", "-99"),
#'   A3a = c("15", "-88")
#' )
#' eavs_missing_status(df)
eavs_missing_status <- function(data) {
  recode_core(data)$status
}

# Shared engine: classify each item column once, return both the numeric
# values and the status labels.
recode_core <- function(data) {
  is_item <- vapply(
    seq_along(data),
    function(i) item_column(data[[i]], names(data)[i]),
    logical(1)
  )
  values <- data
  status <- data
  for (i in which(is_item)) {
    cl <- classify_values(as.character(data[[i]]))
    values[[i]] <- cl$value
    status[[i]] <- factor(cl$status, levels = .eavs_status_levels)
  }
  list(
    values = tibble::as_tibble(values),
    status = tibble::as_tibble(status)
  )
}

# Is a column a numeric EAVS item, as opposed to an identifier or text field?
item_column <- function(x, name) {
  if (any(grepl(paste(.eavs_id_patterns, collapse = "|"), name, perl = TRUE))) {
    return(FALSE)
  }
  if (is.numeric(x)) {
    return(TRUE)
  }
  if (!is.character(x)) {
    return(FALSE)
  }
  trimmed <- trimws(strip_code_label(x))
  keep <- !is.na(x) & trimmed != "" &
    !(tolower(trimmed) %in% names(.eavs_token_status))
  if (!any(keep)) {
    return(FALSE)
  }
  all(!is.na(suppressWarnings(as.numeric(trimmed[keep]))))
}

# Turn a character item column into numeric values and matching status labels.
classify_values <- function(x) {
  x <- strip_code_label(x)
  n <- length(x)
  status <- rep("reported", n)
  value <- rep(NA_real_, n)

  trimmed <- trimws(x)
  blank <- is.na(x) | trimmed == ""
  status[blank] <- "blank"

  norm <- tolower(trimmed)
  tok <- !blank & norm %in% names(.eavs_token_status)
  status[tok] <- unname(.eavs_token_status[norm[tok]])

  rest <- !blank & !tok
  num <- suppressWarnings(as.numeric(trimmed[rest]))
  rstatus <- rep("reported", length(num))
  rvalue <- num

  sent <- match(as.character(num), names(.eavs_sentinels))
  is_sent <- !is.na(sent)
  rstatus[is_sent] <- unname(.eavs_sentinels[sent[is_sent]])
  rvalue[is_sent] <- NA_real_

  neg <- !is.na(num) & num < 0 & !is_sent
  rstatus[neg] <- "other_missing"
  rvalue[neg] <- NA_real_

  bad <- is.na(num)
  rstatus[bad] <- "other_missing"
  rvalue[bad] <- NA_real_

  status[rest] <- rstatus
  value[rest] <- rvalue
  list(value = value, status = status)
}

File: py/src/tidyeavs/recode.py
"""Decoding EAVS missing-value codes.

Mirrors ``r/R/recode.R``. The classification rules are subtle enough that the
two implementations are kept deliberately parallel: identifier and text columns
are left alone, a column counts as a numeric item only when every substantive
value in it parses as a number, and because every EAVS item is a count or a rate
any negative value is read as missing.
"""

from __future__ import annotations

import re

import pandas as pd

from ._constants import ID_PATTERNS, SENTINELS, STATUS_LEVELS, TOKEN_STATUS

# Sentinels matched on the parsed number rather than its text, which makes
# "-88", "-088" and "-88.0" all resolve to the same code, as they do in R.
_SENTINEL_VALUES = {float(code): status for code, status in SENTINELS.items()}

_CODE_LABEL = re.compile(r"^\s*(-?[0-9]+)\s*:.*$")
_ID_RE = tuple(re.compile(pattern, re.IGNORECASE) for pattern in ID_PATTERNS)


def strip_code_label(values: pd.Series) -> pd.Series:
    """Reduce a 2016-style ``"CODE: Label"`` value to its numeric code.

    The 2016 EAVS writes sentinels as ``"-999999: Data Not Available"`` rather
    than a bare number. Plain values pass through unchanged.
    """
    return values.str.replace(_CODE_LABEL, lambda m: m.group(1), regex=True)


def is_item_column(values: pd.Series, name: str) -> bool:
    """Is this a numeric EAVS item, as opposed to an identifier or text field?"""
    if any(pattern.search(name) for pattern in _ID_RE):
        return False
    if pd.api.types.is_numeric_dtype(values):
        return True
    if not (
        pd.api.types.is_string_dtype(values) or pd.api.types.is_object_dtype(values)
    ):
        return False

    text = values.astype("string")
    trimmed = strip_code_label(text).str.strip()
    keep = text.notna() & (trimmed != "") & ~trimmed.str.lower().isin(TOKEN_STATUS)
    if not keep.any():
        return False
    return bool(pd.to_numeric(trimmed[keep], errors="coerce").notna().all())


def _classify(values: pd.Series) -> tuple[pd.Series, pd.Series]:
    """Return (numeric values, status labels) for one item column."""
    text = strip_code_label(values.astype("string"))
    trimmed = text.str.strip()
    lowered = trimmed.str.lower()

    blank = text.isna() | (trimmed == "")
    token = ~blank & lowered.isin(TOKEN_STATUS)
    rest = ~blank & ~token

    number = pd.to_numeric(trimmed.where(rest), errors="coerce")

    status = pd.Series("reported", index=values.index, dtype=object)
    status[blank] = "blank"
    status[token] = lowered[token].map(TOKEN_STATUS)

    sentinel = rest & number.isin(_SENTINEL_VALUES)
    status[sentinel] = number[sentinel].map(_SENTINEL_VALUES)
    # A negative that is not a recognized sentinel is still missing, since no
    # EAVS item is legitimately negative. So is a value that will not parse.
    status[rest & number.notna() & (number < 0) & ~sentinel] = "other_missing"
    status[rest & number.isna()] = "other_missing"

    reported = rest & number.notna() & ~sentinel & (number >= 0)
    value = pd.Series(pd.NA, index=values.index, dtype="Float64")
    value[reported] = number[reported].astype("Float64")

    labels = pd.Categorical(status, categories=STATUS_LEVELS, ordered=True)
    return value, pd.Series(labels, index=values.index)


def _recode_core(data: pd.DataFrame) -> tuple[pd.DataFrame, pd.DataFrame]:
    values = data.copy()
    status = data.copy()
    for name in data.columns:
        if is_item_column(data[name], str(name)):
            values[name], status[name] = _classify(data[name])
    return values, status


def recode_missing(data: pd.DataFrame) -> pd.DataFrame:
    """Convert item columns to numbers, with sentinel values set to ``NA``.

    Identifier and text columns—FIPS codes, jurisdiction and state names,
    comment and write-in fields—are left untouched, so this is safe to call on
    a whole raw year from :func:`tidyeavs.read`.

    Recognized sentinels keep their specific meaning, which
    :func:`missing_status` reports; collapsing everything to ``NA`` would lose
    the distinction between "does not apply" and "data not available".
    """
    return _recode_core(data)[0]


def missing_status(data: pd.DataFrame) -> pd.DataFrame:
    """Report *why* each item value is, or is not, missing.

    Returns a frame the same shape as ``data`` with each item value replaced by
    an ordered categorical: ``reported``, ``does_not_apply``, ``not_available``,
    ``valid_skip``, ``other_missing``, or ``blank``. Identifier and text columns
    are unchanged.

    This is the companion to :func:`recode_missing`, and the reason to reach for
    it is coverage: a state total computed with ``skipna=True`` silently covers
    only the jurisdictions that reported, so it is worth knowing how many did.
    """
    return _recode_core(data)[1]
